File: R/countXO.R
######################################################################
#
# countXO.R
#
# copyright (c) 2008-2013, Karl W Broman
# last modified Sep, 2013
# first written Feb, 2008
#
#     This program is free software; you can redistribute it and/or
#     modify it under the terms of the GNU General Public License,
#     version 3, as published by the Free Software Foundation.
#
#     This program is distributed in the hope that it will be useful,
#     but without any warranty; without even the implied warranty of
#     merchantability or fitness for a particular purpose.  See the GNU
#     General Public License, version 3, for more details.
#
#     A copy of the GNU General Public License, version 3, is available
#     at http://www.r-project.org/Licenses/GPL-3
#
# Part of the R/qtl package
# Contains: countXO
#
######################################################################

######################################################################
#
# countXO: Count number of obligate crossovers for each individual
#          on individual chromosomes or overall
#
# if bychr=TRUE, return matrix with no. obligate crossovers for each
#                individual on each chromosome
#        =FALSE, return vector with total no. crossovers across the
#                selected chromosomes
######################################################################

countXO <-
    function(cross, chr, bychr=FALSE)
{
    if(!any(class(cross) == "cross"))
        stop("Input should have class \"cross\".")

    # pull out relevant chromosome
    if(!missing(chr)) cross <- subset(cross,chr=chr)
    chr.name <- names(cross$geno)

    type <- class(cross)[1]
    n.ind <- nind(cross)
    n.chr <- nchr(cross)
    nxo <- matrix(0, ncol=n.chr, nrow=n.ind)
    id <- getid(cross)
    if(is.null(id)) id <- 1:n.ind
    dimnames(nxo) <- list(id, chr.name)

    for(i in 1:n.chr) {

        chrtype <- class(cross$geno[[i]])
        if(chrtype=="X") xchr <- TRUE
        else xchr <- FALSE

        # which type of cross is this?
        if(type == "f2") {
            if(!xchr) # autosomal
                func <- "R_countXO_f2"
            else func <- "R_countXO_bc"        # X chromsome
        }
        else if(type == "bc" || type=="riself" || type=="risib" || type=="dh" || type=="haploid") func <- "R_countXO_bc"
        else if(type == "4way") func <- "R_countXO_4way"
        else if(type=="ri4self" || type=="ri4sib" || type=="ri8self" || type=="ri8sib" || type=="bgmagic16") {
            func <- "R_countXO_ril48"
            if(xchr)
                warning("countXO not working properly for the X chromosome for 4- or 8-way RIL.")
        }
        else
            stop("ripple not available for cross ", type)

        # data to be input
        genodat <- cross$geno[[i]]$data
        genodat[is.na(genodat)] <- 0
        n.mar <- ncol(genodat)

        if(n.mar > 1) {
            z <- .C(func,
                    as.integer(n.ind),
                    as.integer(n.mar),
                    as.integer(genodat),
                    oblxo=as.integer(rep(0,n.ind)),
                    PACKAGE="qtl")

            nxo[,i] <- z$oblxo
        }
    }

    if(!bychr) nxo <- apply(nxo, 1, sum)

    nxo
}

# end of countXO.R

File: src/countXO.c
/**********************************************************************
 * 
 * countXO.c
 *
 * copyright (c) 2008-9, Karl W Broman
 *
 * last modified Apr, 2009
 * first written Feb, 2008
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License,
 *     version 3, as published by the Free Software Foundation.
 * 
 *     This program is distributed in the hope that it will be useful,
 *     but without any warranty; without even the implied warranty of
 *     merchantability or fitness for a particular purpose.  See the GNU
 *     General Public License, version 3, for more details.
 * 
 *     A copy of the GNU General Public License, version 3, is available
 *     at http://www.r-project.org/Licenses/GPL-3
 *
 * C functions for the R/qtl package
 *
 * These functions are for comparing marker orders by counts of 
 * obligate crossovers
 *
 * Contains: R_countXO_bc, R_countXO_f2, R_countXO_4way, countXO
 *           R_countXO_ril48
 *  
 **********************************************************************/

#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/PrtUtil.h>
#include <R_ext/Applic.h>
#include <R_ext/Utils.h>
#include "util.h"
#include "ripple.h"
#include "countXO.h"

/**********************************************************************
 * 
 * countXO
 *
 * This function counts the number of obligate crossovers for each 
 * individual on a chromosome
 *
 * Input:
 *
 * n_ind    = no. individuals
 *
 * n_mar    = no. markers
 *
 * n_gen    = no. possible genotypes
 *
 * geno     = genotype data [n_ind x n_mar]
 *
 * nxo      = the output; the number of obligate crossovers for each
 *            individual
 *
 * countxo  = function to count the number of obligate crossovers in
 *            an interval and to update the current inferred genotype
 *            (specific for backcross, intercross, and four-way cross)
 *
 **********************************************************************/

void countXO(int n_ind, int n_mar, int n_gen, int *geno,
	     int *nxo, int countxo(int *curgen, int nextgen))
{
  int **Geno;
  int j, k, curgen;

  /* reorganize genotype data and marker order matrix */
  reorg_geno(n_ind, n_mar, geno, &Geno);
  
  for(j=0; j<n_ind; j++) { /* loop over individuals */

    R_CheckUserInterrupt(); /* check for ^C */
    nxo[j] = 0;

    /* genotype at first marker */
    curgen = Geno[0][j];
    for(k=1; k<n_mar; k++) /* loop over markers */
      /* count no obligate crossovers and update current genotype */
      nxo[j] += countxo(&curgen, Geno[k][j]);
  }
}


/**********************************************************************
 * 
 * R_countXO_bc
 *
 * Wrapper for call from R for a backcross
 * 
 **********************************************************************/

void R_countXO_bc(int *n_ind, int *n_mar, int *geno, 
		  int *nxo)
{
  countXO(*n_ind, *n_mar, 2, geno, nxo, countxo_bc);
}




/**********************************************************************
 * 
 * R_countXO_f2
 *
 * Wrapper for call from R for an intercross
 * 
 **********************************************************************/

void R_countXO_f2(int *n_ind, int *n_mar, int *geno, 
		  int *nxo)
{
  countXO(*n_ind, *n_mar, 4, geno, nxo, countxo_f2);
}


/**********************************************************************
 * 
 * R_countXO_4way
 *
 * Wrapper for call from R for a four-way cross
 * 
 **********************************************************************/

void R_countXO_4way(int *n_ind, int *n_mar, int *geno, 
		    int *nxo)
{
  countXO(*n_ind, *n_mar, 4, geno, nxo, countxo_4way);
}

/**********************************************************************
 * 
 * R_countXO_ril48
 *
 * Wrapper for call from R for 4- or 8-way RIL by selfing or sibmating
 * 
 **********************************************************************/

void R_countXO_ril48(int *n_ind, int *n_mar, int *geno, 
		    int *nxo)
{
  countXO(*n_ind, *n_mar, 4, geno, nxo, countxo_ril48);
}

/* end of countXO.c */

